File: ctci/chapter_4_trees_graphs/cpp/singly_linked_list.c
#include "singly_linked_list.h"

/**
 * Node constructor.
 */
Node::Node(TreeNode *value)
{
    this->value = value;
    this->next = NULL;
};

/**
 * Singly linked list constructor.
 */
LinkedList::LinkedList()
{
    this->head = NULL;
    this->length = 0;
}

/**
 * Adds a new node to the head of a singly linked list.
 */
Node *LinkedList::add_node(TreeNode *node)
{
    Node *new_node = new Node(node);
    if (!this->head)
        this->head = new_node;
    else
    {
        new_node->next = this->head;
        this->head = new_node;
    }
    this->length++;
    return new_node;
}

/**
 * Prints all elements of a singly linked list.
 */
void LinkedList::print_list()
{
    Node *curr = this->head;
    while (curr->next)
    {
        cout << curr->value->value << " -> ";
        curr = curr->next;
    }
    cout << curr->value->value << endl;
}

File: ctci/chapter_2_linked_lists/singly_linked_list.py
class Node:

    def __init__(self, value: int):
        """
        Node constructor.
        """
        self.value = value
        self.next = None
    
    def print_node(self):
        """
        Prints node.
        """
        curr = self
        while curr:
            print(curr.value, end="-> " if curr.next else "\n")
            curr = curr.next


class SinglyLinkedList:

    def __init__(self):
        """
        Singly linked list constructor.
        """
        self.head = None
        self.length = 0

    def add_to_front(self, value: int) -> Node:
        """
        Adds a new node to the head of a singly linked list.
        """
        node = Node(value)
        if self.head is None:
            self.head = node
        else:
            node.next = self.head
            self.head = node
        self.length += 1
        return node

    def delete_node(self, index: int):
        """
        Deletes a node at a given index.
        """
        curr = self.head
        if index < 0 or index >= self.length:
            raise IndexError
        # Delete head.
        if index == 0:
            self.head = curr.next
            del(curr)
            return
        # Delete last node.
        if index == self.length - 1:
            while curr.next:
                prev = curr
                curr = curr.next
            prev.next = None
            del(curr)
            return
        while index != 0 and curr:
            prev = curr
            curr = curr.next
            index -= 1
        prev.next = curr.next
        del(curr)

    def print_linked_list(self):
        """
        Prints all elements of a singly linked list.
        """
        curr = self.head
        while curr:
            print(curr.value, end="-> " if curr.next else "\n")
            curr = curr.next
